idx/memory/memory.go

```diff
@@ -213,20 +214,19 @@ func (m *MemoryIdx) Stop() {
 func (m *MemoryIdx) Update(point schema.MetricPoint, partition int32) (idx.Archive, int32, bool) {
 	pre := time.Now()
 
-	m.Lock()
-	defer m.Unlock()
+	m.RLock()
+	defer m.RUnlock()
 
 	existing, ok := m.defById[point.MKey]
 	if ok {
-		oldPart := existing.Partition
 		if LogLevel < 2 {
 			log.Debug("metricDef with id %v already in index", point.MKey)
 		}
 
-		if existing.LastUpdate < int64(point.Time) {
-			existing.LastUpdate = int64(point.Time)
+		if atomic.LoadInt64(&existing.LastUpdate) < int64(point.Time) {
+			atomic.SwapInt64(&existing.LastUpdate, int64(point.Time))
```

Comment:
note that this is racey.
let's say existing.LastUpdate is very old (30 days ago)
then a point comes in for 29 days ago, and concurrently another one for a day ago via a different kafka partition, and then no more points.
in that case, we can have concurrent Update calls, resulting in the LastUpdate field being updated to 29 days ago, but never to a day ago.
note that for any kafka given partition, carbon stream or prometheus POST we never have overlap in our Update calls.

so in practice, doesn't seem like an issue, but perhaps we should document this something under "tradeoffs and extremely rare edge cases" or something.

Reply:
or we can solve it by either:
1) doing CompareAndSwap in a loop until we're able to swap for the value we wanted to swap
2) confirming the swapped out value (return value of SwapInt64) is smaller than what we swapped in. if not, put the old value back, check that we didn't swap for an ever higher value (placed by a concurrent Update call), etc

Reply:
I don't understand this. if you use a write lock, you can lock, check value, if we have a larger one, update, unlock. this works regardless of the order between two concurrent update operations. (the most recent will always survive).

I think my proposal above will also solve it, and at almost no additional cost.

Reply:
I was extending it to the `Partition` update as well. Perhaps `Partition` should only be updated when we have a newer timestamp as well?

Reply:
the partition property is really only to shard the index in cassandra, so nodes on startup know which metrics they're responsible for.
i'm not sure if we even properly support live partition changes of metrics (i.e. whether after the change we properly display both the old and new data)
under concurrent updates it's probably ok for the partition to be ambiguous ("under transition"), but once update operations are serialized, the later one should probably always win, even when data is sent out of order. I think MT's behavior in these scenarios is so undefined that probably either way works.